src/ripple/shamap/impl/SHAMapSync.cpp

```diff
@@ -332,51 +336,56 @@ bool SHAMap::getNodeFat (SHAMapNodeID wanted, std::vector<SHAMapNodeID>& nodeIDs
         return false;
     }
 
-    int count;
-    bool skipNode = false;
-    do
-    {
-
-        if (skipNode)
-            skipNode = false;
-        else
-        {
-            Serializer s;
-            node->addRaw (s, snfWIRE);
-            nodeIDs.push_back (wanted);
-            rawNodes.push_back (std::move (s.peekData ()));
-        }
+    std::stack<std::tuple <SHAMapTreeNode*, SHAMapNodeID, int>> stack;
+    stack.emplace (node, nodeID, depth);
 
-        if ((!fatRoot && wanted.isRoot ()) || node->isLeaf ()) // don't get a fat root_, can't get a fat leaf
-            return true;
+    while (! stack.empty ())
+    {
+        std::tie (node, nodeID, depth) = stack.top ();
+        stack.pop ();
 
-        SHAMapTreeNode* nextNode = nullptr;
-        SHAMapNodeID nextNodeID;
+        // Add this node to the reply
+        Serializer s;
+        node->addRaw (s, snfWIRE);
+        nodeIDs.push_back (nodeID);
+        rawNodes.push_back (std::move (s.peekData()));
 
-        count = 0;
-        for (int i = 0; i < 16; ++i)
+        if (node->isInner())
         {
-            if (!node->isEmptyBranch (i))
+            // We descend inner nodes with only a single child
+            // without decrementing the depth
+            int bc = node->getBranchCount();
+            if ((depth > 0) || (bc == 1))
             {
-                SHAMapNodeID nextNodeID = wanted.getChildNodeID (i);
-                nextNode = descendThrow (node, i);
-                ++count;
-                if (fatLeaves || nextNode->isInner ())
+                // We need to process this node's children
+                for (int i = 0; i < 16; ++i)
```

Comment:
In general I'm against magic numbers but they need to be sufficiently magical - this one is hardly magical. There's a price of admission required to understand rippled code. I think its reasonable that knowing every SHAMap inner node has 16 children is included in the price. Some constants *should* be explained, I just don't feel like this is one of them. FYI there are at least 40 places where `16` is used to indicate the number of children per inner node. I don't see that changing those places to `SHAMapItem::maxChildren` will provide sufficient enlightenment.